src/NHibernate/Type/ComponentType.cs

```diff
@@ -520,6 +520,8 @@ public override object SemiResolve(object value, ISessionImplementor session, ob
 		{
 			//note that this implementation is kinda broken
 			//for components with many-to-one associations
+			if (ReturnedClass.IsInstanceOfType(value))
+				return value;
```

Comment:
`SemiResolve` documentation does not state it has to support "re-entry" (supporting being supplied with an already resolved value), but it is used that way at many places in the code.

In the case of a component, the non-resolved value is an `object[]` array, and the returned class cannot be one, so checking if the value type is already the resolved one.